File: test_demo.cpp
#include <iostream>
#include <string>
#include <random>
#include <ctime>
#include <cctype>

char generate_random_character () {
    // generate a random char
    std::string list_of_random_chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890+-=_?;:";

    std::random_device os_seed;
    uint_least32_t random_number_seed = os_seed();

    std::mt19937 generator(random_number_seed);
    std::uniform_int_distribution< uint_least32_t > distribute(0, list_of_random_chars.length() - 1);

    int random_index = distribute(generator);
    //std::cout << random_index << std::endl;
    char random_char = list_of_random_chars[random_index];
    return random_char;
}

int main () {
    std::string run_status = "yes";
    while (run_status == "yes") {

        // get this user input
        std::cout << "do you want to generate a random password? yes or no?" << std::endl;
        std::string user_input;
        std::cin >> user_input;
        std::cout << user_input << std::endl;
        // check if the input is yes or no
        if (user_input == "yes") {
            std::cout << "a password will now be generated" << std::endl;
            // generate the password
            std::cout << "please enter a password length" << std::endl;
            std::string user_input_password_length;
            std::cin >> user_input_password_length;
            int converted_string = 0;
            try {
                // convert the string to int
                converted_string = std::stoi(user_input_password_length);
                std::cout << "the password length is: " << converted_string << std::endl;
                if (converted_string >= 1) {
                    // generate the appropriate number of random chars
                    std::string generated_password;
                    for (int x = 1; x <= converted_string; x = x + 1) {
                        // generate the random character
                        char random_char = generate_random_character();
                        // append the random char to the password string
                        generated_password = generated_password + random_char;
                    }
                    // display the password
                    std::cout << "your password is:\n" << generated_password << std::endl;
                } else {
                    std::cout << "error: the password must be 1 character long or more" << std::endl;
                }
            }
            catch (...) {
                std::cout << "error: you have not entered a valid number" << std::endl;
            }
        } else if (user_input == "no") {
            std::cout << "the program will exit" << std::endl;
            run_status = "no";
            // exit the program
        } else {
            std::cout << "please enter yes or no" << std::endl;
        }
    }
    return 0;
}

File: test_demo.py
import random

def generate_random_char():
    # set the possible random characters
    list_of_possible_characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
    # get a random index and retrieve the random character
    random_index = random.randint(0,len(list_of_possible_characters) - 1)
    random_char = list_of_possible_characters[random_index]
    return random_char

# main program loop
main_loop_status = "running"
while main_loop_status == "running":
    # get the user input -> yes/no on generating password
    user_input_generate_password = input("do you want to generate a random password? yes/no: ")
    # check if the user input is yes/no
    if (user_input_generate_password == "yes"):
        # generate the password
        print ("a password will now be generated")
        user_input_password_length = input("how many characters must be the password be? Enter a number:")
        # convert the password length user input into a string
        password_length_converted = 0
        try:
            # attempt to convert the password length user input into an int
            password_length_converted = int(user_input_password_length)
            # if the password length is a valid int
            if (password_length_converted <= 0):
                # if the password length is 0
                print ("the password length must be greater than 0")
            if (password_length_converted > 0):
                # if the password length is greater than 0
                print ("a password of ",password_length_converted," length will now be generated", sep='')
                # generate the password
                password_string = ""
                for x in range(1,password_length_converted + 1):
                    # generate the random character
                    random_char = generate_random_char()
                    password_string = password_string + random_char
                print ("your password is: ",password_string,sep='')
        except:
            print ("invalid number. please enter a number without a decimal")
    if (user_input_generate_password == "no"):
        # exit the program
        print ("the program will exit")
        main_loop_status = "not running"
    if (user_input_generate_password != "yes") and (user_input_generate_password != "no"):
        # if the user input is not yes AND not no
        print ("invalid input, please enter yes or no")

